extensions/notification/NotificationQueuePopulator.js

```diff
@@ -119,22 +119,78 @@ class NotificationQueuePopulator extends QueuePopulatorExtension {
+     * to display according to the
+     * versioning state of the object
+     * @param {Object} value log entry object
+     * @param {Object} overheadFields - extra fields missing from the log entry
+     * @return {String} versionId
+     */
+    _getVersionId(value, overheadFields) {
+        const versionId = value.versionId || (overheadFields && overheadFields.versionId);
+        const isNullVersion = value.isNull;
+        const isVersioned = !!versionId;
+        // Versioning suspended objects have
+        // a versionId, however it is internal
+        // and should not be used to get the object
+        if (isNullVersion || !isVersioned) {
+            return null;
+        }
+        return versionId;
+    }
+
+    /**
+     * Decides if we should process the entry.
+     * Since we get both master and version events,
+     * we need to avoid pushing two notifications for
+     * the same event.
+     * - For non versiond buckets, we process the master
+     * objects' events.
+     * - For versioned buckets, we process version events
+     * and ignore all master events.
+     * - For versioning suspended buckets, we need to process
+     * both master and version events, as the master is considered
+     * a separate version.
+     * @param {String} key object key
+     * @param {Object} value object metadata
+     * @return {boolean} - true if entry is valid
+     */
+    _shouldProcessEntry(key, value) {
+        const isMaster = isMasterKey(key);
+        const isVersionedMaster = isMaster && !!value.versionId;
+        const isNullVersion = isMaster && value.isNull;
+        if (!isMaster
+            || !isVersionedMaster
+            || isNullVersion
+            ) {
+            return true;
+        }
+        return false;
+    }
+
     /**
      * Process object entry from the log
      *
      * @param {String} bucket - bucket
      * @param {String} key - object key
      * @param {Object} value - log entry object
      * @param {String} type - entry type
-     * @param {String} commitTimestamp - when the entry was written, used as a fallback
+     * @param {Object} overheadFields - extra context fields missing from the log entry
+     * @param {String} overheadFields.commitTimestamp - when the entry was written, used as a fallback
      *   if no last-modified MD attribute available
+     * @param {String} overheadFields.versionId - version id involved in this operation, if the log entry
+     *   does not contain it
      * @return {undefined}
      */
-    _processObjectEntry(bucket, key, value, type, commitTimestamp) {
-        const versionId = value.versionId || null;
-        if (!isMasterKey(key)) {
+    _processObjectEntry(bucket, key, value, type, overheadFields) {
+        if (!this._shouldProcessEntry(key, value)) {
             return undefined;
         }
+
+        const versionId = this._getVersionId(value, overheadFields);
+        const baseKey = this._extractVersionedBaseKey(key);
```

Comment:
By the way, is `key` the raw key returned from the oplog? If so, we need to strip it of its V1 prefix `\x7fM` or `\x7fV`. But it's not related to this PR, we should maybe have a separate ticket to handle this.